### PR TITLE
#11964 Type annotate more tests

### LINKED ISSUE #11964: type annotate more tests
**Is your feature request related to a problem? Please describe.**
tests are pretty easy to type annotate just lots of manual `-> None:`

**Describe the solution you'd like**
type annotate some more tests

### PR DESCRIPTION
**Pull request by graingert**
## Scope and purpose

Fixes #11964

### CODE REVIEW

File: src/twisted/web/test/_util.py [hunk 1]
```diff
@@ -13,10 +13,13 @@
...
 from twisted.web._flatten import flattenString
 from twisted.web.error import FlattenerError
+from twisted.web.http import Request
+from twisted.web.resource import IResource
 from twisted.web.template import Flattenable
+from .requesthelper import DummyRequest
 
 
-def _render(resource, request):
+def _render(resource: IResource, request: Request | DummyRequest) -> Deferred[None]:
```
Comment (graingert) [inline src/twisted/web/test/_util.py:22]: should there be an Interface that both http.Request and DummyRequest implement? Currently there's no interface with `notifyFinish()`

File: src/twisted/words/test/test_tap.py [hunk 2]
```diff
@@ -24,37 +28,37 @@ def setUp(self):
...
         self._loginTest(opt)
 
-    def test_auth(self):
+    def test_auth(self) -> None:
         """
         Tests that the --auth command generates a checker.
         """
         opt = tap.Options()
         opt.parseOptions(["--auth", "file:" + self.file.name])
         self._loginTest(opt)
```
Comment (graingert) [inline src/twisted/words/test/test_tap.py:59]: should this be?
```suggestion
        return self._loginTest(opt)
```

how is the assertion error in self._loginTest collected if the Deferred isn't returned? I verified that the AssertionError is collected if I change the callback to d.addCallback to raise

File: src/twisted/trial/test/test_doctest.py [hunk 3]
```diff
@@ -26,15 +26,15 @@ def test_id(self):
...
+    def test_basicTrialIntegration(self) -> None:
         """
         L{loadDoctests} loads all of the doctests in the given module.
         """
         loader = runner.TestLoader()
         suite = loader.loadDoctests(mockdoctest)
         self.assertEqual(7, suite.countTestCases())
 
-    def _testRun(self, suite):
+    def _testRun(self, suite: pyunit.TestSuite) -> None:
```
Comment (adiroiban) [inline src/twisted/trial/test/test_doctest.py:37]: just a drive-by comment.

I think that pyunit.TestSuite is ok

I am not sure we still need `twisted.trial.unittest.TestSuite`

File: src/twisted/web/test/_util.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/web/test/_util.py:22]: the HTTP interfaces are not the best .

I think the current annotation is ok. thanks

What I did, I have created my own DummyRequest that  inherits from http.Request and does the minimum overrwriting possible.

File: src/twisted/words/test/test_tap.py (hunk 2 above)
Comment (adiroiban) [reply src/twisted/words/test/test_tap.py:59]: It  should return a deferred... or inlineCallbacks :)

If you have time, please update these tests. thanks

**Review by adiroiban — APPROVED** [review]
Looks good. Many thanks.

I think we can have this merged, and then we can have separate PR to look at the "controversial" code.

Can we have macOS tests with 3.8 enabled in this branch ?

I left a few minor comment inline.

Thanks again.

File: src/twisted/trial/test/test_doctest.py (hunk 3 above)
Comment (graingert) [reply src/twisted/trial/test/test_doctest.py:37]: I checked at runtime and it's a regular pyunit.UnitTest

**Comment by graingert** [comment]
> Can we have macOS tests with 3.8 enabled in this branch ?

I think that still makes sense in a separate PR

**Comment by graingert** [comment]
@adiroiban 

> "controversial" code.

What do you mean by that?

**Comment by adiroiban** [comment]
> > "controversial" code.
> What do you mean by that?

Changes like `request: Request | DummyRequest` or returning a deferred in the test_tap.py

**Comment by graingert** [comment]
Ah I see
